File: java/src/HarnessReact.java
/**
 * Test Harness — Java ReAct Loop
 * Runs a battery of tasks, captures full traces, reports pass/fail/tokens/time.
 *
 * Same deps as ReactLoop.java:
 *   com.squareup.okhttp3:okhttp:4.12.0
 *   com.fasterxml.jackson.core:jackson-databind:2.17.0
 *
 * export ANTHROPIC_API_KEY=sk-ant-...
 * javac -cp okhttp*.jar:jackson*.jar HarnessReact.java ReactLoop.java
 * java  -cp .:okhttp*.jar:jackson*.jar HarnessReact
 */

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.function.BiFunction;
import java.util.stream.Collectors;

public class HarnessReact {

    static final ObjectMapper JSON = new ObjectMapper();

    // ── Types ─────────────────────────────────────────────────────────────────
    record ValidationResult(boolean passed, String reason) {}

    @FunctionalInterface
    interface Validator extends BiFunction<String, ReactLoop.Trace, ValidationResult> {}

    record TestCase(
        String    name,
        String    task,
        Validator validate,
        int       maxIterations,
        String    systemExtra
    ) {
        TestCase(String name, String task, Validator validate) {
            this(name, task, validate, 30, null);
        }
    }

    record TestResult(
        String  name,
        boolean passed,
        String  reason,
        String  answer,
        int     iterations,
        long    totalTokens,
        long    elapsedMs,
        String  error
    ) {}

    // ── Validators ────────────────────────────────────────────────────────────
    static Validator contains(String... keywords) {
        return (answer, trace) -> {
            String lower = answer.toLowerCase();
            for (String k : keywords) {
                if (!lower.contains(k.toLowerCase())) {
                    return new ValidationResult(false, "missing keyword: " + k);
                }
            }
            return new ValidationResult(true, "OK");
        };
    }

    static Validator answerNotEmpty() {
        return (answer, trace) -> answer.isBlank()
            ? new ValidationResult(false, "answer is empty")
            : new ValidationResult(true, "OK");
    }

    static Validator usedTool(String... toolNames) {
        return (answer, trace) -> {
            Set<String> used = trace.steps.stream()
                .filter(s -> "action".equals(s.type()))
                .map(ReactLoop.TraceStep::tool)
                .collect(Collectors.toSet());
            for (String t : toolNames) {
                if (!used.contains(t)) {
                    return new ValidationResult(false, "tool not called: " + t);
                }
            }
            return new ValidationResult(true, "OK");
        };
    }

    static Validator thoughtBeforeAction() {
        return (answer, trace) -> {
            boolean lastWasThought = false;
            for (ReactLoop.TraceStep step : trace.steps) {
                switch (step.type()) {
                    case "thought" -> lastWasThought = true;
                    case "action"  -> {
                        if (!"think".equals(step.tool()) && !"finish".equals(step.tool())) {
                            if (!lastWasThought) {
                                return new ValidationResult(false,
                                    "action '" + step.tool() + "' not preceded by thought");
                            }
                            lastWasThought = false;
                        }
                    }
                }
            }
            return new ValidationResult(true, "OK");
        };
    }

    static Validator minIterations(int n) {
        return (answer, trace) -> trace.iterations >= n
            ? new ValidationResult(true, "OK")
            : new ValidationResult(false, "expected >= " + n + " iterations, got " + trace.iterations);
    }

    static Validator all(Validator... validators) {
        return (answer, trace) -> {
            for (Validator v : validators) {
                ValidationResult r = v.apply(answer, trace);
                if (!r.passed()) return r;
            }
            return new ValidationResult(true, "OK");
        };
    }

    // ── Test suite ────────────────────────────────────────────────────────────
    static List<TestCase> buildSuite() {
        return List.of(
            new TestCase("basic_answer",
                "What is 2 + 2? Use the think tool to reason, then finish.",
                all(contains("4"), thoughtBeforeAction())),

            new TestCase("code_execution",
                "Write and run a bash command to compute the sum of squares from 1 to 10. Report the result.",
                all(contains("385"), usedTool("run_bash"), thoughtBeforeAction())),

            new TestCase("file_write_and_read",
                "Write 'ReAct harness test' to /tmp/harness_test_java.txt, then read it back and confirm the content.",
                all(contains("ReAct harness test"), usedTool("write_file", "read_file"), thoughtBeforeAction())),

            new TestCase("multi_step_reasoning",
                "Calculate the 10th Fibonacci number using bash. Then the 20th. Report the ratio of 20th to 10th.",
                all(contains("55", "6765"), usedTool("run_bash"), minIterations(3), thoughtBeforeAction()),
                20, null),

            new TestCase("error_recovery",
                "Try to read /tmp/no_such_file_java_99.txt. If it fails, create it with 'created by agent', then read it back.",
                all(contains("created by agent"), usedTool("read_file", "write_file"), thoughtBeforeAction())),

            new TestCase("system_prompt_respected",
                "Tell me your name.",
                contains("ARIA"),
                10, "Your name is ARIA. Always introduce yourself as ARIA."),

            new TestCase("finish_called",
                "Say hello and finish.",
                all(answerNotEmpty(), thoughtBeforeAction()),
                10, null)
        );
    }

    // ── Runner ────────────────────────────────────────────────────────────────
    static List<TestResult> runHarness(List<TestCase> suite) throws Exception {
        List<TestResult> results = new ArrayList<>();
        String sep = "=".repeat(64);

        System.out.printf("%n%s%n  ReAct Harness — Java — %d tests%n%s%n%n", sep, suite.size(), sep);

        for (TestCase tc : suite) {
            System.out.printf("▶  %-40s ... ", tc.name());
            long startMs = System.currentTimeMillis();

            TestResult result;
            try {
                ReactLoop.ReactResult rr = ReactLoop.runReact(
                    tc.task(),
                    tc.systemExtra(),
                    tc.maxIterations(),
                    null
                );

                ValidationResult vr = tc.validate().apply(rr.answer, rr.trace);
                result = new TestResult(
                    tc.name(), vr.passed(), vr.reason(),
                    rr.answer,
                    rr.trace.iterations,
                    rr.trace.totalTokens,
                    rr.trace.elapsedMs,
                    ""
                );
            } catch (Exception e) {
                result = new TestResult(
                    tc.name(), false, "Exception", "",
                    0, 0, System.currentTimeMillis() - startMs,
                    e.getMessage()
                );
            }

            String status = result.passed() ? "✅ PASS" : "❌ FAIL";
            System.out.printf("%s  (%dms, %d tok, %d iter)%n",
                status, result.elapsedMs(), result.totalTokens(), result.iterations());

            if (!result.passed()) {
                System.out.printf("   Reason : %s%n", result.reason());
                if (!result.error().isEmpty())
                    System.out.printf("   Error  : %s%n", truncate(result.error(), 300));
                if (!result.answer().isEmpty())
                    System.out.printf("   Answer : %s%n", truncate(result.answer(), 200));
            }

            results.add(result);
        }

        long passed      = results.stream().filter(TestResult::passed).count();
        long totalTokens = results.stream().mapToLong(TestResult::totalTokens).sum();
        long totalMs     = results.stream().mapToLong(TestResult::elapsedMs).sum();

        System.out.printf("%n%s%n  Results : %d/%d passed%n  Tokens  : %d%n  Time    : %.2fs%n%s%n%n",
            sep, passed, results.size(), totalTokens, totalMs / 1000.0, sep);

        // Write JSON report
        String reportPath = "/tmp/react_harness_report_java.json";
        ArrayNode arr = JSON.createArrayNode();
        for (TestResult r : results) {
            var n = JSON.createObjectNode();
            n.put("name",          r.name());
            n.put("passed",        r.passed());
            n.put("reason",        r.reason());
            n.put("answer",        truncate(r.answer(), 500));
            n.put("iterations",    r.iterations());
            n.put("total_tokens",  r.totalTokens());
            n.put("elapsed_ms",    r.elapsedMs());
            n.put("error",         truncate(r.error(), 500));
            arr.add(n);
        }
        Files.writeString(Path.of(reportPath), JSON.writerWithDefaultPrettyPrinter().writeValueAsString(arr));
        System.out.printf("  Report  → %s%n", reportPath);

        return results;
    }

    static String truncate(String s, int n) {
        return s == null ? "" : s.length() <= n ? s : s.substring(0, n) + "...";
    }

    // ── Entry point ───────────────────────────────────────────────────────────
    public static void main(String[] args) throws Exception {
        List<TestResult> results = runHarness(buildSuite());
        boolean anyFailed = results.stream().anyMatch(r -> !r.passed());
        System.exit(anyFailed ? 1 : 0);
    }
}

File: typescript/src/harness_react.ts
/**
 * Test Harness — TypeScript ReAct Loop
 * Runs a battery of tasks, captures full traces, reports pass/fail/tokens/time.
 *
 * npm install @anthropic-ai/sdk
 * export ANTHROPIC_API_KEY=sk-ant-...
 * npx ts-node harness_react.ts
 */

import * as fs from "fs";
import { runReact, Trace, Step } from "./react_loop";

// ── Types ─────────────────────────────────────────────────────────────────────
type Validator = (answer: string, trace: Trace) => { passed: boolean; reason: string };

interface TestCase {
  name:          string;
  task:          string;
  validate:      Validator;
  maxIterations?: number;
  system?:       string;
}

interface TestResult {
  name:         string;
  passed:       boolean;
  reason:       string;
  answer:       string;
  iterations:   number;
  totalTokens:  number;
  elapsedMs:    number;
  error:        string;
}

// ── Validators ────────────────────────────────────────────────────────────────
function contains(...keywords: string[]): Validator {
  return (answer) => {
    const missing = keywords.filter(k => !answer.toLowerCase().includes(k.toLowerCase()));
    return missing.length === 0
      ? { passed: true,  reason: "OK" }
      : { passed: false, reason: `Missing keywords: ${missing}` };
  };
}

function answerNotEmpty(): Validator {
  return (answer) => answer.trim()
    ? { passed: true, reason: "OK" }
    : { passed: false, reason: "Answer is empty" };
}

function usedTool(...toolNames: string[]): Validator {
  return (_, trace) => {
    const used = new Set(
      trace.steps.filter(s => s.type === "action").map(s => (s as any).tool)
    );
    const missing = toolNames.filter(t => !used.has(t));
    return missing.length === 0
      ? { passed: true,  reason: "OK" }
      : { passed: false, reason: `Expected tool calls not found: ${missing}` };
  };
}

function thoughtBeforeAction(): Validator {
  return (_, trace) => {
    let lastWasThought = false;
    for (const step of trace.steps) {
      if (step.type === "thought") {
        lastWasThought = true;
      } else if (step.type === "action") {
        const tool = (step as any).tool;
        if (tool !== "think" && tool !== "finish") {
          if (!lastWasThought) {
            return { passed: false, reason: `Action '${tool}' not preceded by a Thought` };
          }
          lastWasThought = false;
        }
      }
    }
    return { passed: true, reason: "OK" };
  };
}

function minIterations(n: number): Validator {
  return (_, trace) => trace.iterations >= n
    ? { passed: true,  reason: "OK" }
    : { passed: false, reason: `Expected >= ${n} iterations, got ${trace.iterations}` };
}

function all(...validators: Validator[]): Validator {
  return (answer, trace) => {
    for (const v of validators) {
      const result = v(answer, trace);
      if (!result.passed) return result;
    }
    return { passed: true, reason: "OK" };
  };
}

// ── Test suite ────────────────────────────────────────────────────────────────
const TEST_SUITE: TestCase[] = [
  {
    name:     "basic_answer",
    task:     "What is 2 + 2? Use the think tool to reason, then finish.",
    validate: all(contains("4"), thoughtBeforeAction()),
  },
  {
    name:     "code_execution",
    task:     "Write and run Node.js code to compute the sum of squares from 1 to 10. Report the result.",
    validate: all(contains("385"), usedTool("run_node"), thoughtBeforeAction()),
  },
  {
    name:     "file_write_and_read",
    task:     "Write the string 'ReAct harness test' to /tmp/harness_test_ts.txt, then read it back and confirm.",
    validate: all(contains("ReAct harness test"), usedTool("write_file", "read_file"), thoughtBeforeAction()),
  },
  {
    name:     "multi_step_reasoning",
    task:     "Calculate the 10th Fibonacci number using code. Then the 20th. Report the ratio of 20th to 10th.",
    validate: all(contains("55"), contains("6765"), usedTool("run_node"), minIterations(3), thoughtBeforeAction()),
  },
  {
    name:     "error_recovery",
    task:     "Try to read /tmp/no_such_file_ts_99.txt. If it fails, create it with 'created by agent', then read it back.",
    validate: all(contains("created by agent"), usedTool("read_file", "write_file"), thoughtBeforeAction()),
  },
  {
    name:     "system_prompt_respected",
    task:     "Tell me your name.",
    system:   "Your name is ARIA. Always introduce yourself as ARIA.",
    validate: contains("ARIA"),
  },
  {
    name:     "finish_called",
    task:     "Say hello and finish.",
    validate: all(answerNotEmpty(), thoughtBeforeAction()),
  },
];

// ── Runner ────────────────────────────────────────────────────────────────────
async function runHarness(suite: TestCase[] = TEST_SUITE): Promise<TestResult[]> {
  const results: TestResult[] = [];

  console.log(`\n${"=".repeat(64)}`);
  console.log(`  ReAct Harness — TypeScript — ${suite.length} tests`);
  console.log(`${"=".repeat(64)}\n`);

  for (const tc of suite) {
    process.stdout.write(`▶  ${tc.name} ... `);
    const t0 = Date.now();
    const result: TestResult = {
      name: tc.name, passed: false, reason: "",
      answer: "", iterations: 0, totalTokens: 0, elapsedMs: 0, error: "",
    };

    try {
      const { answer, trace } = await runReact(tc.task, {
        systemExtra:   tc.system,
        maxIterations: tc.maxIterations ?? 30,
      });

      result.answer      = answer;
      result.iterations  = trace.iterations;
      result.totalTokens = trace.totalTokens;
      result.elapsedMs   = trace.elapsedMs;

      const { passed, reason } = tc.validate(answer, trace);
      result.passed = passed;
      result.reason = reason;

    } catch (e: unknown) {
      result.passed  = false;
      result.reason  = "Exception";
      result.error   = (e as Error).message ?? String(e);
      result.elapsedMs = Date.now() - t0;
    }

    const status = result.passed ? "✅ PASS" : "❌ FAIL";
    console.log(`${status}  (${result.elapsedMs}ms, ${result.totalTokens} tok, ${result.iterations} iter)`);

    if (!result.passed) {
      console.log(`   Reason : ${result.reason}`);
      if (result.error)  console.log(`   Error  : ${result.error.slice(0, 300)}`);
      if (result.answer) console.log(`   Answer : ${result.answer.slice(0, 200)}`);
    }

    results.push(result);
  }

  const passed      = results.filter(r => r.passed).length;
  const totalTokens = results.reduce((s, r) => s + r.totalTokens, 0);
  const totalMs     = results.reduce((s, r) => s + r.elapsedMs, 0);

  console.log(`\n${"=".repeat(64)}`);
  console.log(`  Results : ${passed}/${results.length} passed`);
  console.log(`  Tokens  : ${totalTokens}`);
  console.log(`  Time    : ${(totalMs / 1000).toFixed(2)}s`);
  console.log(`${"=".repeat(64)}\n`);

  const reportPath = "/tmp/react_harness_report_ts.json";
  fs.writeFileSync(reportPath, JSON.stringify(results.map(r => ({
    ...r, answer: r.answer.slice(0, 500), error: r.error.slice(0, 500),
  })), null, 2));
  console.log(`  Report  → ${reportPath}`);

  return results;
}

if (require.main === module) {
  runHarness().then(results => {
    const failed = results.filter(r => !r.passed);
    process.exit(failed.length === 0 ? 0 : 1);
  }).catch(e => {
    console.error("FATAL:", e);
    process.exit(1);
  });
}
